src/CoreWCF.Primitives/tests/OperationContextTests.cs

```diff
@@ -0,0 +1,147 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Http;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using CoreWCF.Configuration;
+using Microsoft.AspNetCore;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.Extensions.DependencyInjection;
+using Xunit;
+
+namespace CoreWCF.Primitives.Tests
+{
+    public partial class OperationContextTests
+    {
+        private static readonly Uri _serviceAddress = new Uri($"http://localhost:8080/Test", UriKind.Absolute);
+
+        [ServiceContract]
+        private interface IContract
+        {
+            [OperationContract]
+            void Operation();
+        }
+
+        // IncludeExceptionDetailInFaults is not needed to reproduce the original bug,
+        // it's here purely for better Exception visibility in case of a failure.
+        [ServiceBehavior(IncludeExceptionDetailInFaults = true)]
+        private class Service : IContract
+        {
+            public void Operation()
+            {
+                var context = OperationContext.Current;
+
+                // This sleep is not needed to reproduce the original bug in a real environment,
+                // but is essential to reproduce it in a test environment.
+                Thread.Sleep(100);
```

Comment:
I believe this test is going to be fragile (if the issue wasn't fixed). My understanding of why you sleep is to allow time for another thread to modify the same Holder and the delay is to allow time for that to happen. I believe there's an easier more deterministic way to achieve this and you should be able to do it with a single request. Modify the service to look like this:
```c#
        [ServiceBehavior(IncludeExceptionDetailInFaults = true)]
        public class Service : IContract
        {
            public ManualResetEventSlim RequestStarted { get; } = new ManualResetEventSlim(false);
            public ManualResetEventSlim ContextModified { get; } = new ManualResetEventSlim(false);
            public void Operation()
            {
                var context = OperationContext.Current;
                RequestStarted.Set();
                ContextModified.Wait(TimeSpan.FromSeconds(10));

                if (context != OperationContext.Current)
                {
                    throw new FaultException("OperationContext.Current was replaced in mid-operation.");
                }
            }
        }
```
In the `Startup.ConfigureServicesMethod` add the `Service` to DI as a singleton.
```c#
        private class Startup
        {
            public void ConfigureServices(IServiceCollection services)
            {
                services.AddServiceModelServices();
                services.AddSingleton<Service>();
            }
            /// Rest of class
         }
```
In the test method, you start the request but don't await it yet. You then wait on the `RequestStarted` event, modify the `OperationContext.Current`, then set the `ContextModified` event, then await the request.
```c#
    var responseTask = client.SendAsync(request);
    var service = host.Services.GetRequiredService<Service>();
    service.RequestStarted.Wait(TimeSpan.FromSeconds(10));
    OperationContext.Current = null;
    service.ContextModified.Set();
    var response = await responseTask;
    var responseBody = await response.Content.ReadAsStringAsync();
```

You only need to make one request so remove the Enumerable.Range which kicks off multiple parallel requests and just make a single request. Basically this works by kicking off the http request but not waiting for the response yet, then waiting on a `ManualResetEventSlim` to know that the request has started and the initial `OperationContext.Current` value has been captured. The service implementation is now waiting to be signaled that the `OperationContext.Current` has been modified. So the test sets it to null, which will be reflected in the service call, then signals the service to continue with the operation where it then picks up the null value.  

As a general rule don't use Sleep calls to try to trigger a bug in a probabilistic way. The CI test hosts tend to only have 2 CPU's configured so tests like yours are likely to fail there as there's a good chance that with 100 runs there will be at least one where not two calls are executing concurrently (which is what's required to hit this bug).

Reply:
I have pushed my implementation a perfectly synchronized approach for the test, based on your comments.
However beware that the timeouts make the test fragile in the CI environment. It was failed with a timeout of 10 seconds, but passed with a timeout of 30 seconds. I don't really know what is the correct number to put there.

Reply:
Being fragile due to taking too long is different than being fragile due to racing code. Put the timeout at 30 seconds and if that turns out to be too low, I can increase it later.

Reply:
The current pushed code already has the timeout at 30 seconds.

Reply:
Great, I'll merge this. Thanks for doing this and contributing a repro test. I plan to do a release including this fix next week unless anything else comes up which derails things.